### ISSUE TITLE
Versioned API objects should be in a self-contained package

### ISSUE THREAD

**Issue by fabioy** [kubernetes/kubernetes#7584]
TL;DR: The versioned API types should be self-contained package. For example, the file "pkg/api/v1beta3/types.go" is not self contained and contains tendrils to other internal types (see below). This is somewhat cruddy for other clients wanting to import minimal source from k8s just to call its APIs. It'd be nice to have a single package to import to get the API object definitions (i.e. for json unmarshalling, etc).

Observed outside of package depencies in "pkg/api/v1beta3/types.go":

"pkg/runtime" - The "List" type contains a field "Items []runtime.RawExtension". "RawExtension" is a trivial struct containing only a "[]byte".

"pkg/types" - Various places uses the "types.UID" struct, which is an alias to "string".

"pkg/util" - Two types from this package are used: "util.Time" and "util.IntOrString". The former is a wrapper on Go's Time object, with extra functions for custom JSON serialization (RFC3339 compliant) in "pkg/util/time.go". "IntOrString" is the dual-mode struct holding an int value or a string ("pkg/util/util.go").

"pkg/api/resource" - The "ResourceList" object is a map whose keys are "resource.Quantity" objects. The "Quantity" object contains a field to "inf.Dec", which is an infinite precision decimal package imported from "speter.net/go/exp/math/dec/inf". This is the only package which is external to k8s.

I know that some of the json annotations added are to support swagger, but the go->swagger->go thing seems like an unnecessary complexity.

**Comment by bgrant0607** [comment]
Related: #7111, #2306, #1451, #1744, #6210, #4874, #2742, #4851

**Comment by bgrant0607** [comment]
I don't think we have any go->swagger->go paths yet.

**Comment by bgrant0607** [comment]
See https://github.com/technosophos/kubelite
cc @lavalamp

**Comment by lavalamp** [comment]
This is covered in my current proposal: #16062

**Comment by lavalamp** [comment]
We should definitely offer something like that, but a version of it that won't bitrot.

**Comment by lavalamp** [comment]
This is closely related to #4874 and #5660.

**Comment by bgrant0607** [comment]
This came up last week. I think we agreed that making it easy for third parties to consume just the API types would be useful. Here's an example of someone doing that:
https://twitter.com/ae6rt/status/792124301268824064

The main question was whether we want the versioned client and apiserver to both import the same types, or whether they should have separate copies. The concern about the latter is that it would cause code duplication of libraries using the types. The concern about the former is leakage of apiserver implementation details (e.g., additional attributes specified via tags and/or comments) and the need of stability of the types for the client library.

cc @mml @caesarxuchao @kubernetes/sig-api-machinery

**Comment by smarterclayton** [comment]
The server API work will clean up the spread of some of these types.  Much of the refactoring here is on my list of things we must do for internal sanity:

1. Eliminate helpers that use internal apis - part of Chao's work for controllers
2. Make internal a subordinate package to external (internal can directly import external safely if necessary) to prevent internal "leaks"
3. Unify versioned common types in a single versioned package - server API 
4. Runtime is not imported from versioned or internal apis

I still think server and client code should reuse the same objects - those types are our IDL today, and as long as they are we benefit from from acting as if they are idl than not.  I'm not aware of leaks today that aren't in the list above.  

Client stability is an interesting question but I don't think paying the cost of splitting the API types is a must have for moving in the direction of refining external types.

**Comment by caesarxuchao** [comment]
I prefer client code and server reusing the same types. For these reasons:
1. like @bgrant0607 said, there are utility functions used by both server and client. Reusing types eliminates duplicates. This also makes migrating to client-go easier.
2. it makes field level versioning (https://github.com/kubernetes/kubernetes/issues/34508) easier. Otherwise server needs to track the versioning of client types.

client-go follows semver. So far we makes breaking changes to client-go more frequently than breaking changes in API, so I think client stability is less concerning.

> Make internal a subordinate package to external (internal can directly import external safely if necessary) to prevent internal "leaks"

@smarterclayton do you mean internal can import external but not the other way around?

**Comment by smarterclayton** [comment]
Conversion is very expensive.  Most of our conversions are actually between
identical fields.  We've got a (mostly) complete implementation that uses
the unsafe package to do "safe" transformations as allowed by unsafe.
However, over time I'd like to make internal be less important outside of
the API server.  With your work to move many clients to versioned, internal
will be mostly geared towards conversion and API server specific action.
I'd like to (in the near term) have internal be structured in a way that
indicates that internal types are mostly the same as particular external
types, except in specific cases.

Imagine, for instance:

```
package internal

import "k8s.io/kubernetes/pkg/api/v1"

// PodStatus is exactly the same as the external PodStatus
type PodStatus v1.PodStatus
```

This makes it much clearer that internal exists as a convenience for
conversion.  I imagine that we might move versioned conversions into their
own packages that only internal can see.  That more accurately reflects the
intent of internal - hub for conversion, and convenience for central code
that must be version independent.

**Comment by bgrant0607** [comment]
We've also discussed eliminating the internal rep. We could just use the latest versioned rep instead. We mostly deliberately don't use a more convenient internal rep, because it would make conversion more manual and error-prone.

Elsewhere we've discussed clients that need to support multiple versions, such as HPA, though I wonder whether the frequency of that could be reduced by longer periods of overlapping support for both versions in apiserver.

**Comment by bgrant0607** [comment]
Most utility functions should be used by clients, not apiserver. 

One case for multiple versions is in tests.

**Comment by smarterclayton** [comment]
This has mostly been resolved by refactorings in 1.6.  We have a cleaner versioned / internal separation.  Closing unless there's something important I missed.
